x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AbstractAggregationDataExtractor.java

```diff
@@ -0,0 +1,167 @@
+    implements DataExtractor {
+
+    private static final Logger LOGGER = LogManager.getLogger(AbstractAggregationDataExtractor.class);
+
+    /**
+     * The number of key-value pairs written in each batch to process.
+     * This has to be a number that is small enough to allow for responsive
+     * cancelling and big enough to not cause overhead by calling the
+     * post data action too often. The value of 1000 was determined via
+     * such testing.
+     */
+    private static int BATCH_KEY_VALUE_PAIRS = 1000;
+
+    protected final Client client;
+    protected final AggregationDataExtractorContext context;
+    private boolean hasNext;
+    private boolean isCancelled;
+    private AggregationToJsonProcessor aggregationToJsonProcessor;
+    private ByteArrayOutputStream outputStream;
+
+    AbstractAggregationDataExtractor(Client client, AggregationDataExtractorContext dataExtractorContext) {
+        this.client = Objects.requireNonNull(client);
+        context = Objects.requireNonNull(dataExtractorContext);
+        hasNext = true;
+        isCancelled = false;
+        outputStream = new ByteArrayOutputStream();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return hasNext;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return isCancelled;
+    }
+
+    @Override
+    public void cancel() {
+        LOGGER.debug("[{}] Data extractor received cancel request", context.jobId);
+        isCancelled = true;
+        hasNext = false;
+    }
+
+    @Override
+    public Optional<InputStream> next() throws IOException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        if (aggregationToJsonProcessor == null) {
+            Aggregations aggs = search();
+            if (aggs == null) {
+                hasNext = false;
+                return Optional.empty();
+            }
+            initAggregationProcessor(aggs);
+        }
+
+        return Optional.ofNullable(processNextBatch());
+    }
+
+    private Aggregations search() throws IOException {
+        LOGGER.debug("[{}] Executing aggregated search", context.jobId);
+        SearchResponse searchResponse = executeSearchRequest(buildSearchRequest(buildBaseSearchSource()));
+        LOGGER.debug("[{}] Search response was obtained", context.jobId);
+        ExtractorUtils.checkSearchWasSuccessful(context.jobId, searchResponse);
+        return validateAggs(searchResponse.getAggregations());
+    }
+
+    private void initAggregationProcessor(Aggregations aggs) throws IOException {
+        aggregationToJsonProcessor = new AggregationToJsonProcessor(context.timeField, context.fields, context.includeDocCount,
+            context.start);
+        aggregationToJsonProcessor.process(aggs);
+    }
+
+    protected SearchResponse executeSearchRequest(T searchRequestBuilder) {
+        return ClientHelper.executeWithHeaders(context.headers, ClientHelper.ML_ORIGIN, client, searchRequestBuilder::get);
```

Comment:
This will fail if a rollup search has been selected but the user who created the datafeed doesn't have permission to use rollups.

We have to decide how we want to handle this situation better.  The two options are:

1. If the index pattern includes a rollup index when the datafeed is created and the user does not have permission to run rollup searches then refuse to create the datafeed
2. During the check on whether to use a normal search or rollup search, if the user doesn't have permission to run rollup searches then silently fall back to a normal search

I think option 1 is probably better because it avoids the possibility of a datafeed that the user thinks will use rollups silently not working as expected.

To implement that, look in `TransportPutDatafeedAction.masterOperation`.  It currently does a `HasPrivilegesRequest` to check that the user can search the desired indices.  If the index pattern provided includes a rollup index then it should _also_ check privileges for rollup search (normal search privilege is still required as well because rollup search uses that internally).

There is still a theoretical flaw in this as an index pattern could be provided that does not match a rollup index at the time the datafeed is created, but does later on.  For example, suppose a datafeed is created against `farequote*`.  Initially this matches `farequote-20181029`, `farequote-20181030`, `farequote-20181031`, etc.  But later it also matches `farequote_rollup`.  With the current logic the datafeed would switch to rollup search at that time, and start to fail if the user did not have permission to use rollup search.  But this is an edge case and I'm happy to just cover the more obvious case initially.

Reply:
@droberts195 how is this behavior any different than when a user creates a datafeed in which they don't have access to that specific index?

The reason for why I have it the way it is, is because of behavior parity with how we handle permission issues when reading data from an index. 

I will happily change to verify the user has admin privileges for the indices. I would just note that the Rollup folks are working towards simply requiring regular search permissions for `_rollup_search` so that it is treated as a simple index read, which would give it parity with other datafeed read actions.

Reply:
> how is this behavior any different than when a user creates a datafeed in which they don't have access to that specific index?

The problem is that there are many types of "access".  Each action can be allowed or disallowed for a particular user.  There are [index privileges](https://github.com/elastic/elasticsearch/blob/1ce3c92a2d4d2c387011f47628de346da2413271/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java#L44) that group together multiple actions.  But there's also some overlap in what actions the different privileges allow.

> the Rollup folks are working towards simply requiring regular search permissions for `_rollup_search`

They'll do this by adjusting the actions that are allowed by the [`READ_AUTOMATON`](https://github.com/elastic/elasticsearch/blob/1ce3c92a2d4d2c387011f47628de346da2413271/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java#L44).  But our [privileges check in `TransportPutDatafeedAction.java`](https://github.com/elastic/elasticsearch/blob/20c6c9c542835f1620a495c54f2d84235f338fa7/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDatafeedAction.java#L93) is checking specifically for the ability to run the `_search` action, not for the "read" index privilege.  I think we should also check for the ability to use the `_rollup_search` action at that same point in the code _if_ the index pattern provided includes a rollup index (which will require an extra test to be done early in `TransportPutDatafeedAction`).

Reply:
I see what you are saying @droberts195 I was thinking of the extraction layer, not the creation layer of the datafeed. Will update shortly